subsys/bluetooth/mesh/adv_ext.c

```diff
@@ -0,0 +1,290 @@
+	if (err) {
+		BT_ERR("Failed setting adv data: %d", err);
+		atomic_clear_bit(adv.flags, ADV_FLAG_ACTIVE);
+		return err;
+	}
+
+	adv.timestamp = k_uptime_get();
+
+	err = bt_le_ext_adv_start(adv.instance, start);
+	if (err) {
+		BT_ERR("Advertising failed: err %d", err);
+		atomic_clear_bit(adv.flags, ADV_FLAG_ACTIVE);
+	}
+
+	return err;
+}
+
+static int buf_send(struct net_buf *buf)
+{
+	struct bt_le_ext_adv_start_param start = {
+		.num_events =
+			BT_MESH_TRANSMIT_COUNT(BT_MESH_ADV(buf)->xmit) + 1,
+	};
+	uint16_t duration, adv_int;
+	struct bt_data ad;
+	int err;
+
+	adv_int = MAX(ADV_INT_FAST_MS,
+		      BT_MESH_TRANSMIT_INT(BT_MESH_ADV(buf)->xmit));
+	/* Upper boundary estimate: */
+	duration = start.num_events * (adv_int + 10);
+
+	BT_DBG("type %u len %u: %s", BT_MESH_ADV(buf)->type,
+	       buf->len, bt_hex(buf->data, buf->len));
+	BT_DBG("count %u interval %ums duration %ums",
+	       BT_MESH_TRANSMIT_COUNT(BT_MESH_ADV(buf)->xmit) + 1, adv_int,
+	       duration);
+
+	ad.type = bt_mesh_adv_type[BT_MESH_ADV(buf)->type];
+	ad.data_len = buf->len;
+	ad.data = buf->data;
+
+	/* Only update advertising parameters if they're different */
+	if (adv_param.interval_min != BT_MESH_ADV_SCAN_UNIT(adv_int)) {
+		adv_param.interval_min = BT_MESH_ADV_SCAN_UNIT(adv_int);
+		adv_param.interval_max = adv_param.interval_min;
+		atomic_set_bit(adv.flags, ADV_FLAG_UPDATE_PARAMS);
+	}
+
+	adv.cb = BT_MESH_ADV(buf)->cb;
+	adv.cb_data = BT_MESH_ADV(buf)->cb_data;
+
+	err = adv_start(&adv_param, &start, &ad, 1, NULL, 0);
+	net_buf_unref(buf);
+	adv_send_start(duration, err);
+
+	return err;
+}
+
+static void send_pending_adv(struct k_work *work)
+{
+	struct net_buf *buf;
+	int err;
+
+	atomic_clear_bit(adv.flags, ADV_FLAG_SCHEDULED);
+
+	while ((buf = net_buf_get(&bt_mesh_adv_queue, K_NO_WAIT))) {
+		/* busy == 0 means this was canceled */
+		if (!BT_MESH_ADV(buf)->busy) {
+			net_buf_unref(buf);
+			continue;
+		}
+
+		BT_MESH_ADV(buf)->busy = 0U;
+		err = buf_send(buf);
+		if (!err) {
+			return; /* Wait for advertising to finish */
+		}
+	}
+
+	/* No more pending buffers */
+	if (IS_ENABLED(CONFIG_BT_MESH_PROXY)) {
+		BT_DBG("Proxy Advertising");
+		err = bt_mesh_proxy_adv_start();
+		if (!err) {
+			atomic_set_bit(adv.flags, ADV_FLAG_PROXY);
+		}
+	}
+}
+
+static void schedule_send(void)
+{
+	uint64_t timestamp = adv.timestamp;
+	int64_t delta;
+
+	if (atomic_test_and_clear_bit(adv.flags, ADV_FLAG_PROXY)) {
+		bt_le_ext_adv_stop(adv.instance);
+		atomic_clear_bit(adv.flags, ADV_FLAG_ACTIVE);
+	}
+
+	if (atomic_test_bit(adv.flags, ADV_FLAG_ACTIVE) ||
+	    atomic_test_and_set_bit(adv.flags, ADV_FLAG_SCHEDULED)) {
+		return;
+	}
+
+	/* The controller will send the next advertisement immediately.
+	 * Introduce a delay here to avoid sending the next mesh packet closer
+	 * to the previous packet than what's permitted by the specification.
+	 */
+	delta = k_uptime_delta(&timestamp);
+	k_delayed_work_submit(&adv.work, K_MSEC(ADV_INT_FAST_MS - delta));
```

Comment:
As a general rule, calls should be immediately, but adv.cb ->end may point to a callback in the application layer, causing the task to switch or block, resulting in a delta exceeding the ADV_INT_FAST_MS.

Reply:
Or, assume only single one adv message in the queue, then `adv_sent` will be called, and update `timestamp`,  then sequence call `schedule_send`, and update `timestamp` again, less than 20 milliseconds later, `send_pending_adv` will call, and send proxy if proxy support and enabled.

If the upper layer sends a mesh message after more than 20 milliseconds(after queue empty), schedule will be called `schedule_send` In this case, `delta` will overflow?

System Power on logic is the same?

Reply:
If the callbacks take several milliseconds, I think that's just something we'll have to accept. Ultimately, we're running in a thread here, and there's nothing we can do to prevent the user from blocking the CPU for tens of milliseconds with a cooperative thread. The main point is to not send two packets on air with less than a 20 millisecond interval, as that would violate the core spec. 

> If the upper layer sends a mesh message after more than 20 milliseconds(after queue empty), schedule will be called schedule_send In this case, delta will overflow?

This is prevented by the `K_MSEC` macro, which forces a non-negative time value with `MAX(0, value)`